### DISCUSSION TITLE
Few enhancements we are currently adding when running on internal devcontainer and container setup.

### DISCUSSION THREAD

**Discussion by mharj** [craftcms/cms#9027] [Ideas]
### resourceBasePath from env
general.php
```php
        // use ENV variable to resourceBasePath if have any
        'resourceBasePath' => App::env('CPRESOURCES_PATH') ?: '@webroot/cpresources',
```

### CRAFT_STORAGE_PATH env
web/index.php
```php
use craft\helpers\App;
...
...
// setup storage path or use default
define('CRAFT_STORAGE_PATH', App::env('CRAFT_STORAGE_PATH') ?: CRAFT_BASE_PATH.'/storage');
// Load and run Craft
define('CRAFT_ENVIRONMENT', App::env('ENVIRONMENT') ?: 'production');
```

This gives us more control where actual IO operations are heading (i.e. caching on own docker volume).

**Comment by brandonkelly** [comment]
Not sure about the `CPRESOURCES_PATH` environment variable, since changing the `resourceBasePath` config setting likely requires changing `resourceBaseUrl` as well.

As for adding support for `CRAFT_STORAGE_PATH`, etc., environment variables, that’s something we could change from Craft’s bootstrap script, so it’s checking for environment variables in addition to constants here:

https://github.com/craftcms/cms/blob/06a2e1d78b8b097f824efede5a6b36807a5bed5a/bootstrap/bootstrap.php#L26-L29

Could be a breaking change as there’s a chance people are already defining those environment variables and not expecting Craft to pay attention to them, so added to the 4.0 milestone.

**Comment by mharj** [comment]
We do already use this with /cpresources and /storage paths which are in temp volume outside of current devcontainer (as main devcontainer might get mapped from Windows volume which is very slow) and I personally haven't seen any issues with this yet. Though prod/qa setups have been using defaults anyway. Quickly checked also Apache setup and we do not alias expose /cpresources directly, so is this kind of broken if resourceBaseUrl is not valid and/or exposed from Apache?
